### PR TITLE
$character attributes command

### PR DESCRIPTION
**Pull request by n1tr0-5urf3r**
Added $character attributes command to calculate char attributes from:
- equipment (static and selectable stats)
- Jewel slots
- Runes
- Infusion
- Base Level

### CODE REVIEW

File: guildwars2/characters.py [hunk 1]
```diff
@@ -256,6 +257,229 @@ def handle_duplicates(upgrades):
...
+                        attribute_name = re.sub('^.* ', '', bonus)
+                        if attribute_name in attr_dict:
+                            attr_dict[attribute_name] += int(modifier)
+                    elif pattern_percentage.match(bonus):
+                        modifier = re.sub(' .*$', '', bonus)
+                        modifier = re.sub('\+', '', modifier)
+                        attribute_name = re.sub(' Duration', 'Duration', bonus)
+                        attribute_name = re.sub('^.* ', '', attribute_name)
+                        if attribute_name in attr_dict:
+                            attr_dict[attribute_name] += int(modifier)
```
Comment (Maselkov) [inline guildwars2/characters.py]: Raises a ValueError since modifier can be a value with a percentage - eg `"5%"`

File: guildwars2/characters.py [hunk 2]
```diff
@@ -459,3 +683,43 @@ def get_crafting(self, character):
...
+        for increase, lvl in lvl_dict.items():
+            if lvl[0] <= level <= lvl[1]:
+                if level < 11:
+                    return increase
+                elif level % 2 == 0:
+                    return increase
+                else:
+                    return 0
+
+    def calcBaselvl(self, level: int, acc_baselvl: int, lvl_dict):
```
Comment (Maselkov) [inline guildwars2/characters.py]: Style: use snake_case for function names

File: guildwars2/characters.py [hunk 3]
```diff
@@ -256,6 +257,229 @@ def handle_duplicates(upgrades):
...
+        # Calculate base health
+        attr_dict["Health"] = self.calcBaseHealth(level, 0,
+                                                  profession_group[profession])
+        attr_dict["Health"] += attr_dict["Vitality"] * 10
+
+        ordered_list = ('Power', 'Toughness', 'defense', 'Vitality', 'Health',
+                        'Precision', 'Critical Chance', 'Ferocity',
+                        'CritDamage', 'ConditionDamage', 'Healing',
+                        'Expertise', 'ConditionDuration', 'Concentration',
+                        'BoonDuration', 'AgonyResistance')
```
Comment (Maselkov) [inline guildwars2/characters.py]: Should probably use some re.sub to insert a space before capital letters, and use str.title() to capitalize properly for better display

File: guildwars2/characters.py [hunk 4]
```diff
@@ -256,6 +257,229 @@ def handle_duplicates(upgrades):
...
+        attr_dict["Precision"] += basevalue
+        # Mapping for old attribute names
+        attr_dict["Concentration"] += attr_dict["BoonDuration"]
+        attr_dict["Ferocity"] += attr_dict["CritDamage"]
+        attr_dict["Expertise"] += attr_dict["ConditionDuration"]
+        # Calculate derivative attributes
+        # Reset to default after mapped to new attribute name
+        attr_dict["CritDamage"] = 150 + round(attr_dict["Ferocity"] / 15, 2)
+        attr_dict["BoonDuration"] = round(attr_dict["Concentration"] / 15, 2)
+        attr_dict["ConditionDuration"] = round(attr_dict["Expertise"] / 15, 2)
```
Comment (Maselkov) [inline guildwars2/characters.py]: These result in a float, which when at 0 looks like `0.0`, while other values are simply `0`

File: guildwars2/characters.py (hunk 1 above)
Comment (n1tr0-5urf3r) [reply guildwars2/characters.py]: Removed % with re.sub

File: guildwars2/characters.py (hunk 3 above)
Comment (n1tr0-5urf3r) [reply guildwars2/characters.py]: Added .capitalize()
Added regex for substitution

File: guildwars2/characters.py (hunk 2 above)
Comment (n1tr0-5urf3r) [reply guildwars2/characters.py]: Refactored calc_base_health() and calc_base_lvl()

File: guildwars2/characters.py (hunk 4 above)
Comment (n1tr0-5urf3r) [reply guildwars2/characters.py]: Added cast to int if 0.0

File: guildwars2/characters.py [hunk 5]
```diff
@@ -459,3 +727,43 @@ def get_crafting(self, character):
...
+                level, lvl_dict)
+            new_lvl = level - 1
+            return self.calc_base_lvl(new_lvl, new_acc, lvl_dict)
+
+    def search_lvl_to_health(self, level: int, health_dict):
+        for increase, lvl in health_dict.items():
+            if lvl[0] <= level <= lvl[1]:
+                return increase
+
+    def calc_base_health(self, level: int, acc_baselvl: int, health_dict):
```
Comment (Maselkov) [inline guildwars2/characters.py]: All those helper methods could be internal functions to main command function

File: guildwars2/characters.py [hunk 6]
```diff
@@ -256,6 +257,273 @@ def handle_duplicates(upgrades):
...
+        await ctx.trigger_typing()
+        try:
+            results = await self.get_character(ctx, character)
+        except APINotFound:
+            return await ctx.send("Invalid character name")
+        except APIError as e:
+            return await self.error_handler(ctx, e)
+        profession = results["profession"].lower()
+        color = self.gamedata["professions"][profession]["color"]
+        icon = self.gamedata["professions"][profession]["icon"]
```
Comment (Maselkov) [inline guildwars2/characters.py]: Use get_profession() added in a recent commit

Comment (n1tr0-5urf3r) [reply guildwars2/characters.py]: Adapted to new get_profession()

File: guildwars2/characters.py (hunk 5 above)
Comment (n1tr0-5urf3r) [reply guildwars2/characters.py]: Moved helper functions inside function

**Pull request merged**